crates/ruff_python_semantic/src/model.rs

```diff
@@ -257,6 +257,20 @@ impl<'a> SemanticModel<'a> {
             .is_some_and(|binding| binding.kind.is_builtin())
     }
 
+    /// Return `true` if `member` is a reference to `builtins.$target`,
+    /// i.e. either `object` (where `object` is not overridden in the global scope),
+    /// or `builtins.object` (where `builtins` is imported as a module at the top level)
+    pub fn match_builtin_expr(&self, expr: &Expr, symbol: &str) -> bool {
+        debug_assert!(!symbol.contains('.'));
+        if let Expr::Name(ast::ExprName { id, .. }) = expr {
+            return id == symbol && self.is_builtin(symbol);
+        }
+        self.seen_module(Modules::BUILTINS)
+            && self
+                .resolve_qualified_name(expr)
+                .is_some_and(|qualified_name| qualified_name.segments() == ["builtins", symbol])
```

Comment:
As written I believe this method will do two lookups in the symbol table in some cases (`self.is_builtin` performs one lookup, and `self. resolve_qualified_name` performs another).

I think you could modify it to look something like this:

```rust
if self.seen_module(Modules::BUILTINS) || expr.as_name_expr().is_some_and(|name| name.id == symbol) {
    self
        .resolve_qualified_name(expr)
         // This next line is pseudo-code.
        .is_some_and(|qualified_name| qualified_name.segments() == ["" | "builtins", symbol])
} else {
  false
}

Reply:
> As written I believe this method will do two lookups in the symbol table in some cases (`self.is_builtin` performs one lookup, and `self. resolve_qualified_name` performs another).

Hmm, will it? Currently:
- If it's an `Expr::Name`, we take the fast path:
  - we return `true` if the `id` of the `Name` matches the symbol and `self.is_builtin(symbol)` is also true
  - else, we return `false`
- Otherwise, in the slow path, we don't actually call `self.is_builtin(symbol)` at all, since as you say it's redundant if we have to call `resolve_qualified_name()`

Reply:
Oh, then the function logic is incorrect, right? Since a name node could be imported from builtins, as in `from builtins import bool`?

Reply:
Correct. Fixed in https://github.com/astral-sh/ruff/pull/10919/commits/f11bf9807199ac286098b4d7a07af17044aaf3c6 :)